**Convolutional Neural Network**

Team members: LMV N1

*   Rocío Sánchez Solórzano 2043109
*   Karen Melissa Morales Moya 2043909
*   Hugo Fernando Nuncio Álvarez 2104134
*   Juan Gerardo Torres Flores 2074382


# Load libraries

In [2]:
import numpy as np
from tensorflow.keras.datasets import mnist
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Conv2D,MaxPooling2D,Flatten,Dense

# Load data

In [3]:
(train_data, train_labels), (test_data, test_labels) = mnist.load_data()

print('train data shape:',train_data.shape)
print('test data shape:',test_data.shape)

11490434/11490434 ━━━━━━━━━━━━━━━━━━━━ 0s 0us/step
train data shape: (60000, 28, 28)
test data shape: (10000, 28, 28)


# Data preprocessing

In [14]:
#normalize pixel values
train_data = train_data/255.0
test_data = test_data/255.0

#respahe the data
train_data = train_data.reshape(-1,28,28,1)
test_data = test_data.reshape(-1,28,28,1)

print('processed train shape:',train_data.shape)

processed train shape: (60000, 28, 28, 1)


# Model selection and training

In [18]:
model = Sequential([
    Conv2D(32, (3,3), activation='relu', input_shape=(28,28,1)),
    MaxPooling2D((2,2)),

    Conv2D(64, (3,3), activation='relu'),
    MaxPooling2D((2,2)),

    Flatten(),

    Dense(64, activation='relu'),
    Dense(10, activation='softmax')
])

In [19]:
#compile
model.compile(
    optimizer='adam',
    loss='sparse_categorical_crossentropy',
    metrics=['accuracy']
)

In [23]:
#train
model.fit(train_data,train_labels,epochs=8,batch_size=32)

Epoch 1/8
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 52s 27ms/step - accuracy: 0.9750 - loss: 0.0819
Epoch 2/8
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 50s 27ms/step - accuracy: 0.9769 - loss: 0.0740
Epoch 3/8
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 50s 26ms/step - accuracy: 0.9786 - loss: 0.0664
Epoch 4/8
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 84s 27ms/step - accuracy: 0.9808 - loss: 0.0621
Epoch 5/8
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 80s 26ms/step - accuracy: 0.9818 - loss: 0.0575
Epoch 6/8
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 51s 27ms/step - accuracy: 0.9831 - loss: 0.0542
Epoch 7/8
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 51s 27ms/step - accuracy: 0.9844 - loss: 0.0510
Epoch 8/8
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 81s 27ms/step - accuracy: 0.9853 - loss: 0.0479


# Model testing

In [24]:
loss,accuracy = model.evaluate(test_data,test_labels)
print('test accuracy:',accuracy)

313/313 ━━━━━━━━━━━━━━━━━━━━ 3s 9ms/step - accuracy: 0.9832 - loss: 0.0505
test accuracy: 0.9832000136375427


In [25]:
#perform five predictions
predictions = model.predict(test_data)

print('\nsample predictions:\n')

for i in range(5):
    prediction = np.argmax(predictions[i])
    actual = test_labels[i]

    print(f'sample {i+1} :')
    print('predicted:', prediction)
    print('actual   :', actual)
    print('-'*20)

313/313 ━━━━━━━━━━━━━━━━━━━━ 3s 8ms/step

sample predictions:

sample 1 :
predicted: 7
actual   : 7
--------------------
sample 2 :
predicted: 2
actual   : 2
--------------------
sample 3 :
predicted: 1
actual   : 1
--------------------
sample 4 :
predicted: 0
actual   : 0
--------------------
sample 5 :
predicted: 4
actual   : 4
--------------------


# Answer the following:

* How many epochs did you use to improve the performance of the model compared with the feedforward neural network of assignment 6?

I used 8 epochs to train the CNN. In contrast, the feedforward model from assignment 5 required 39 iterations.

* Did you get a better performance on this assignment or on assignment 5? Why?

I obtained better performance in this assignment compared to assignment 5 with the feedforward neural network. This improvement is mainly due to the fact that CNNs are specifically designed to work with image data. Unlike feedforward networks, which require flattening the images into one-dimensional vectors, CNNs preserve the two-dimensional structure of the images and are able to detect important spatial features such as edges, shapes, and textures through convolutional layers. This allows the model to learn more meaningful patterns from the data. As a result, the CNN achieves higher accuracy, 98.32%, while the feedforward neural network reaches slightly lower performance, 98.05%, because it loses spatial information during preprocessing.